### PR TITLE
Implemented rest.CategoriesProvider for network policies

### LINKED ISSUE #106176: Network policy not listed
### What happened?

`kubectl get all` outputs all resources except network policy resource

### What did you expect to happen?

i expect it to return all resources including network policy resource

### How can we reproduce it (as minimally and precisely as possible)?

Note: below yaml is taken from: https://kubernetes.io/docs/concepts/services-networking/network-policies/#default-allow-all-ingress-traffic

Step 1:
```yaml
cat <<EOF | kubectl apply -f -
 ---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-all-ingress
spec:
  podSelector: {}
  ingress:
  - {}
  policyTypes:
  - Ingress
EOF
```

Step 2. `kubectl get all`

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.0", GitCommit:"af46c47ce925f4c4ad5cc8d1fca46c7b77d13b38", GitTreeState:"clean", BuildDate:"2020-12-08T17:59:43Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.0", GitCommit:"af46c47ce925f4c4ad5cc8d1fca46c7b77d13b38", GitTreeState:"clean", BuildDate:"2020-12-08T17:51:19Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
none
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.5 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.5 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
$ uname -a
Linux controlplane 5.4.0-1056-gcp #60~18.04.1-Ubuntu SMP Tue Oct 19 15:05:21 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by NikhilSharmaWe**
#### What type of PR is this?
Enhancement 

#### What this PR does / why we need it:
It implements rest.CategoriesProvider for network policies, to also get and group network policies in output of kubectl get all.
#### Which issue(s) this PR fixes:

Fixes #106176

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by NikhilSharmaWe** [comment]
@soltysh implemented rest.CategoriesProvider for
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/networking/networkpolicy/storage/storage.go as you guided. 
But I am getting difficulty in knowing in what other files the same should be implemented. Are these the correct files in which rest.CategoriesProvider should also be implemented. 
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/networking/ingressclass/storage/storage.go
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/networking/ingressclass/storage/storage.go

**Comment by NikhilSharmaWe** [comment]
/kind/bug

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

File: pkg/registry/networking/networkpolicy/storage/storage.go [hunk 1]
```diff
@@ -61,3 +61,11 @@ var _ rest.ShortNamesProvider = &REST{}
 func (r *REST) ShortNames() []string {
 	return []string{"netpol"}
 }
+
+//Implement CategoriesProvider
+var _ rest.CategoriesProvider = &Rest{}
```
Comment (aojea) [inline pkg/registry/networking/networkpolicy/storage/storage.go:66]:
```suggestion
var _ rest.CategoriesProvider = &REST{}
```

File: pkg/registry/networking/networkpolicy/storage/storage.go [hunk 2]
```diff
@@ -61,3 +61,11 @@ var _ rest.ShortNamesProvider = &REST{}
...
+
+// Categories implements the CategoriesProvider interface. Returns a list of categories a resource is part of.
+func (r *REST) Categories() []string{
+	return []string("networkpolicies")
```
Comment (aojea) [inline pkg/registry/networking/networkpolicy/storage/storage.go:70]:
```suggestion
	return []string{"all"}
```

@soltysh is `all` here?

**Review by aojea — CHANGES_REQUESTED** [review]

**Comment by aojea** [comment]
/lgtm cancel
there are some errors and I'm not really sure what value should go in https://github.com/kubernetes/kubernetes/pull/106339#discussion_r747895927

File: pkg/registry/networking/networkpolicy/storage/storage.go (hunk 2 above)
Comment (NikhilSharmaWe) [reply pkg/registry/networking/networkpolicy/storage/storage.go:70]: in this [comment](https://github.com/kubernetes/kubernetes/issues/106176#issuecomment-963088769) @soltysh mentioned that to use `all` is wrong.

**Comment by NikhilSharmaWe** [comment]
@soltysh implemented rest.CategoriesProvider for
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/networking/networkpolicy/storage/storage.go as you guided, is it correct how I did that. 
If the implementation is correct in this case, could you please inform me that in which files the same should be implemented. Are these the correct files in which rest.CategoriesProvider should also be implemented :
* https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/networking/ingressclass/storage/storage.go
* https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/networking/ingress/storage/storage.go

File: pkg/registry/networking/networkpolicy/storage/storage.go (hunk 2 above)
Comment (aojea) [reply pkg/registry/networking/networkpolicy/storage/storage.go:70]: what he says, but you have the format wrong, you need to use brackets not parenthesis
```
return []string{"networkpolicies"}
```

Comment (thockin) [reply pkg/registry/networking/networkpolicy/storage/storage.go:70]: good catch!

Comment (thockin) [reply pkg/registry/networking/networkpolicy/storage/storage.go:70]: I am a little confused.  That comment says "don't add it to 'all'" which I understand.  What's the point of adding a "networkpolicies" category?  If we want a category, we want more than one resource in it, right?

**Review by thockin — COMMENTED** [review]
Unless I misunderstand, this does not solve #106176 (working as intended) and doesn't actually proivide any value  -  why do we want to we do this change?

**Comment by NikhilSharmaWe** [comment]
> Unless I misunderstand, this does not solve #106176 (working as intended) and doesn't actually proivide any value - why do we want to we do this change?

@thockin I am just trying similar to what was done in this [PR](https://github.com/kubernetes/kubernetes/pull/95603) for adding api-extensions in k8s apiserver, I am also not sure about this. Could you please give me some direction about how the similar can be implemented for adding network policies in this case.

**Comment by NikhilSharmaWe** [comment]
> Unless I misunderstand, this does not solve #106176 (working as intended) and doesn't actually proivide any value - why do we want to we do this change?

@thockin I am trying similar to what was done in this [PR](https://github.com/kubernetes/kubernetes/pull/95603) for adding api-extensions in k8s apiserver, I am also not sure about this. Could you please give me some direction about how the similar can be implemented for adding network policies in this case.

**Comment by thockin** [comment]
> Could you please give me some direction about how the similar can be implemented for adding network policies in this case.

What problem are you trying to solve?  I can see how this is hypothetically neat, but without knowing WHY we care, I can't say HOW to solve it.

"network policies" is a single API for now, so this is pointless.  Later we'll have CNP,  so then 2 APIs.  Is the goal to stop there? 

Or is the goal to have a "get network-stuff" call which includes Service, Ingress, Gateway, NetPol,  etc?

TL;DR: Let's not so this unless we know why we need it, and if we feel we know why, let's start with a KEP?

**Comment by NikhilSharmaWe** [comment]
The issue this PR is trying to solve is that `kubectl get` all outputs all resources except network policy resource. So, @soltysh suggestion was to not add the resource in `all` because it can be different for different clusters, instead define something like networks or something similar which will allow you to group all network releated resources, this is the [comment](https://github.com/kubernetes/kubernetes/issues/106176#issuecomment-963088769).

**Comment by NikhilSharmaWe** [comment]
@thockin The issue this PR is trying to solve is that `kubectl get all` outputs all resources except network policy resource. So, @soltysh suggestion was to not add the resource in `all` because it can be different for different clusters, instead define something like networks or something similar which will allow you to group all network releated resources, this is the [comment](https://github.com/kubernetes/kubernetes/issues/106176#issuecomment-963088769).

**Comment by NikhilSharmaWe** [comment]
@thockin Issue this PR is trying to solve is that `kubectl get all` outputs all resources except network policy resource. So, @soltysh suggestion was to not add the resource in `all` because it can be different for different clusters, instead define something like networks or something similar which will allow you to group all network releated resources, this is the [comment](https://github.com/kubernetes/kubernetes/issues/106176#issuecomment-963088769).

**Comment by thockin** [comment]
a) "all" is not really "all" and we should have aa clear convention about adding to it - @soltysh is that written down?

b) I understand wanting to have a group but I don't know what problem that would solve.  AFAIK there is not a "get everything" command.  If we add a network-specific one what problem is it solving?  I have not personally hear anyone ask for this.  That doesn't mean it's not real, just that I don't know it.

**Comment by NikhilSharmaWe** [comment]
@soltysh could you please guide me about what should be changed or how to proceed in this PR.

**Comment by thockin** [comment]
I'm going to close this for now.  If we want to discuss it, let's make a small KEP and discuss there
